src/librbd/ManagedLock.h

```diff
@@ -0,0 +1,184 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MANAGED_LOCK_H
+#define CEPH_LIBRBD_MANAGED_LOCK_H
+
+#include "include/int_types.h"
+#include "include/Context.h"
+#include "include/rados/librados.hpp"
+#include "cls/lock/cls_lock_types.h"
+#include "common/Mutex.h"
+#include <list>
+#include <string>
+#include <utility>
+
+class ContextWQ;
+
+namespace librbd {
+
+template <typename Watcher>
```

Comment:
Sorry, didn't realize it wasn't templated. Looking at this again, the only reason you appear to pass the ```Watcher``` is for the assert that needs to be removed and to get the watch handle on acquire/reacquire.  You already need to do some refactoring to support reacquire, so if you go w/ inheritance to ```ManagedLock```, you could add a pure virtual method to retrieve the watch handle (via ```ExclusiveLock```) -- or ```ExclusiveLock``` could call a set function to provide the handle prior to acquire / reacquire.